### ISSUE TITLE
Support DDP communication hook for speeding up training

### ISSUE THREAD

**Issue by shuyingsunshine21** [PyTorchLightning/pytorch-lightning#6727]
### Motivation

https://pytorch.org/docs/1.8.0/ddp_comm_hooks.html control communicate gradients across workers for `all_reduce` in `DistributedDataParallel`. such as `fp16_compress_hook` converts gradients to fp16 before all reduce is an effective way to improve training speed when using multi nodes.

### Pitch

In `DDPPlugin`, provide an option for specifying `ddp_comm_hook`, and register in `configure_ddp` https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/plugins/training_type/ddp.py#L227
```
    def configure_ddp(self):
        self.pre_configure_ddp()
        self._model = DistributedDataParallel(
            LightningDistributedModule(self.model),
            device_ids=self.determine_ddp_device_ids(),
            **self._ddp_kwargs,
        )
        register_ddp_comm_hook(
            model=self._model,
            is_single_process_single_device=self._is_single_process_single_device,
            ddp_comm_state=self._ddp_comm_state,
            ddp_comm_hook=self._ddp_comm_hook,
            ddp_comm_wrapper=self.ddp_comm_wrapper,
        )
```

**Comment by shuyingsunshine21** [comment]
cc @ananthsub

**Comment by ananthsub** [comment]
Great feature request! Would you like to send out a draft PR for it too 😄 This would be a nice addition to this section: https://pytorch-lightning.readthedocs.io/en/latest/benchmarking/performance.html#when-using-ddp-set-find-unused-parameters-false

fyi @SeanNaren @tchaton @awaelchli

**Comment by shuyingsunshine21** [comment]
sure thing! working on the draft PR, will send out soon

**Comment by tchaton** [comment]
Awesome feature.

I started to add it back then in this PR: https://github.com/PyTorchLightning/pytorch-lightning/pull/5359

https://github.com/PyTorchLightning/pytorch-lightning/pull/5359/files#diff-b21c43e47fc3d5e7d9c8b6f325b85995089672f56639b1e3b3594082478d1841R36

But it is totally out-dated now :)

Best,
T.C
